provider/datagen/src/databake.rs

```diff
@@ -153,213 +205,246 @@ impl DataExporter for BakedDataExporter {
+        .map_err(|_| {
+            DataError::custom("Key component is not a valid Rust identifier").with_key(key)
+        })?;
+
+        let mut path = PathBuf::new();
+        for level in &module_path.segments {
+            self.mod_files
+                .lock()
+                .expect("poison")
+                .entry(path.clone())
+                .or_default()
+                .insert(level.ident.to_string());
+            path = path.join(level.ident.to_string());
+        }
+
+        let struct_type = if is_datetime_skeletons {
+            quote! {
+                &'static [(
+                    &'static [::icu_datetime::fields::Field],
+                    ::icu_datetime::pattern::runtime::PatternPlurals<'static>
+                )]
             }
+        } else {
+            quote! { <#marker as ::icu_provider::DataMarker>::Yokeable }
+        };
 
-            statics.sort_unstable_by(|(a, _), (b, _)| a.cmp(b));
+        let mut map = BTreeMap::new();
+        let mut statics = BTreeMap::new();
 
-            let statics = statics
-                .into_iter()
-                .map(|(ident_string, payload_bake_string)| {
-                    let ident = ident_string.parse::<TokenStream>().unwrap();
-                    let payload_bake = payload_bake_string.parse::<TokenStream>().unwrap();
-                    quote! { static #ident: &DataStruct = &#payload_bake; }
-                });
+        let raw = self
+            .data
+            .lock()
+            .expect("poison")
+            .remove(&key)
+            .unwrap_or_default();
 
-            all_locales.sort_unstable_by(|(a, _), (b, _)| a.cmp(b));
-            let data = all_locales.into_iter().map(|(locale, ident_string)| {
-                let ident = ident_string.parse::<TokenStream>().unwrap();
-                quote! { (#locale, #ident) }
-            });
-
-            // Replace non-ident-allowed tokens. This can still fail if a segment starts with
-            // a token that is not allowed in an initial position.
-            let module_path = syn::parse_str::<syn::Path>(
-                &key.path()
-                    .to_ascii_lowercase()
-                    .replace('@', "_v")
-                    .replace('/', "::"),
-            )
-            .map_err(|_| {
-                DataError::custom("Key component is not a valid Rust identifier").with_key(key)
-            })?;
+        for (payload_bake_string, locales) in raw {
+            let file_name = locales.iter().min().unwrap();
+            let ident =
+                syn::parse_str::<syn::Ident>(&file_name.to_ascii_uppercase().replace('-', "_"))
+                    .unwrap();
+            self.write_to_file(
+                &path.join(file_name),
+                payload_bake_string.parse().unwrap(),
+                true,
+            )?;
+            let file_name = format!("{file_name}.rs.data");
+            let statik = quote! { static #ident: DataStruct = include!(#file_name); };
```

Comment:
suggestion: worth adding a comment to each line saying `// used for locale "foo" and "bar"` or something

Reply:
I think the combined names are a better solution but Shane didn't like that.

Reply:
I wouldn't mind each file having comments stating which locales it is used for, or comments in the mod.rs file. I think it's not really necessary though.